### PR TITLE
perf(common): Optimise TakeRandom with a partial Fisher-Yates shuffle

### LINKED ISSUE #252: perf(common): Optimise TakeRandom with a partial Fisher-Yates shuffle
## Context

Raised by automated reviewers on PR #251 (fix for #246). Out of scope for that correctness fix.

`EnumerableExtensions.TakeRandom<T>` builds an index list and calls `indexes.RemoveAt(randomIndex)` per drawn item. `List<T>.RemoveAt` is O(N), so the overall complexity is O(count × N).

## Suggested resolution

Replace the removal-based selection with a **partial Fisher-Yates shuffle** over the index array (swap the chosen index to the end of the active range and shrink the range), which is O(N) allocation + O(count) selection with no removals, preserving uniform selection probability.

Note: `OrderBy(_ => random.Next())` was also suggested on the PR but should be avoided — sorting by a random key is O(N log N), and re-evaluating non-stable keys can bias or break the sort contract.

While touching this method, also consider whether a negative `count` should throw `ArgumentOutOfRangeException` instead of returning an empty sequence (current behaviour matches LINQ `Take` semantics; changing it would be a breaking behaviour change, so it needs a deliberate decision and a change-log entry if adopted).

Refs: #246

### PR DESCRIPTION
**Pull request by kploch**
## Summary

Implements #252 — replaces `TakeRandom`'s removal-based selection with a partial Fisher-Yates shuffle, and settles the open semantics question from the issue.

## Changes

- `EnumerableExtensions.TakeRandom<T>`: selection now runs a partial Fisher-Yates shuffle directly on a private copy of the source list (swap the chosen element into the front of the active range) — O(n) setup + O(count) selection, replacing the previous O(count × n) `List<T>.RemoveAt` approach. A full-size draw returns the shuffled copy without an extra prefix copy. Uniform selection probability and no-duplicate-picks behaviour preserved.
- XML docs expanded: the `<remarks>` describe the algorithm, and the `count` parameter now documents the clamping (count > size → all items) and zero/negative behaviour.
- Tests: added `TakeRandom_should_return_empty_sequence_when_count_is_negative` (pins the semantics decision), `TakeRandom_should_not_select_the_same_item_twice` (20 × full-size draws must each be a permutation of the source) and `TakeRandom_should_throw_ArgumentNullException_when_source_is_null` (three counts; pins eager validation). The existing reachability (#246 regression), clamping, and edge-case tests guard the refactor.
- `source` is validated (`NotNull`) before the non-positive-count short-circuit, so a null source always throws `ArgumentNullException` — matching `Enumerable.Take`'s eager validation and the pre-optimisation behaviour (review finding).
- Change-log entry added.

## Design Decisions

- **Negative `count` keeps LINQ `Take` semantics** (empty sequence, no throw) — the deliberate decision requested by the issue, confirmed with the repo owner. Consistency with the BCL beats a stricter contract here, and it avoids a breaking change on the eve of v4.0. The behaviour is now documented and pinned by a test rather than incidental.
- Front-swap partial Fisher-Yates (swap chosen into position `i`, draw from `[i, n)`) rather than the swap-to-end variant — identical properties, and the drawn prefix directly yields the result order.
- `OrderBy(_ => random.Next())` was rejected per the issue's own note (O(n log n), unstable-key sort-contract risk).

## Testing

- 45/45 `EnumerableExtensionsTests` pass on net8.0 and net10.0.
- Full solution test suite: zero failures across all 15 assemblies.
- Build: zero new warnings (the only warning on this branch is the pre-existing WebUI IsPackable one, fixed by #261 in this batch).

## Related

Closes #252

## Summary by Sourcery

Optimise EnumerableExtensions.TakeRandom to use a partial Fisher–Yates shuffle with clarified semantics for count and updated tests and change-log entry.

Bug Fixes:
- Define and enforce negative and zero count semantics for TakeRandom to return an empty sequence, matching LINQ Take behavior.

Enhancements:
- Replace removal-based random selection in TakeRandom with a partial Fisher–Yates shuffle over an index array for better performance and preserved uniformity and uniqueness.
- Expand XML documentation for TakeRandom to describe the algorithm and document clamping and count behavior.

Documentation:
- Add a change-log entry documenting the TakeRandom optimization and its clarified count behavior.

Tests:
- Add tests covering negative count returning an empty sequence and ensuring TakeRandom does not select duplicate items in a full-size draw.

___

## **CodeAnt-AI Description**
Speed up random item selection while preserving predictable selection behavior

### What Changed
- `TakeRandom` selects random items without duplicates using a partial shuffle, avoiding repeated removals as the requested count grows
- Requests for zero or negative counts return an empty sequence without reading the source
- Requests larger than the source return all items; full-size selections avoid an unnecessary extra copy
- Selection behavior and edge cases are covered by regression tests

### Impact
`✅ Faster large random selections`
`✅ Lower allocation for full-source selections`
`✅ Predictable empty results for non-positive counts`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved random selection to consistently return unique items.
  * Negative or zero selection counts now return an empty result without errors.
  * Requests larger than the available collection are safely capped.

* **Documentation**
  * Added release documentation describing the updated random selection behavior.

* **Tests**
  * Added coverage for non-positive counts and duplicate prevention.

### CODE REVIEW

File: src/Common/Collections/EnumerableExtensions.cs [hunk 1]
```diff
@@ -187,23 +194,28 @@ public static IEnumerable<TValue> TakeRandom<TValue>(this IEnumerable<TValue> so
     {
         var list = source.ToList();
 
-        var result = new List<TValue>();
+        if (count > list.Count)
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Collections/EnumerableExtensions.cs]: Accepted — 6f02d1b moves the count <= 0 check before ToList(), so a non-positive count returns without enumerating the source (matching Enumerable.Take, which also produces an empty result without eager enumeration). The XML remarks now state this explicitly.

File: src/Common/Collections/EnumerableExtensions.cs [hunk 2]
```diff
@@ -187,23 +194,28 @@ public static IEnumerable<TValue> TakeRandom<TValue>(this IEnumerable<TValue> so
...
         {
-            var indexesItemNum = ThreadSafeRandom.Shared.Next(indexes.Count);
-            var itemIndex = indexes[indexesItemNum];
-
-            result.Add(list[itemIndex]);
-            indexes.RemoveAt(indexesItemNum);
+            var swapIndex = ThreadSafeRandom.Shared.Next(i, indexes.Length);
+            (indexes[i], indexes[swapIndex]) = (indexes[swapIndex], indexes[i]);
+            result.Add(list[indexes[i]]);
         }
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Collections/EnumerableExtensions.cs]: Accepted — 6f02d1b performs the partial swap directly on the source copy (which TakeRandom owns, so mutating it is safe) and returns list.GetRange(0, count). Both the indexes array and the result list are gone, taking allocations from O(2N + K) to O(N). All 42 EnumerableExtensions tests (including the uniformity/reachability and distinctness regression tests) pass on net8.0 and net10.0.

File: src/Common/Collections/EnumerableExtensions.cs [hunk 3]
```diff
@@ -176,37 +176,42 @@ public static IEnumerable<TValue> Shuffle<TValue>(this IEnumerable<TValue> sourc
...
-            var itemIndex = indexes[indexesItemNum];
-
-            result.Add(list[itemIndex]);
-            indexes.RemoveAt(indexesItemNum);
+            var swapIndex = ThreadSafeRandom.Shared.Next(i, list.Count);
+            (list[i], list[swapIndex]) = (list[swapIndex], list[i]);
         }
 
-        return result;
+        return list.GetRange(0, count);
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Collections/EnumerableExtensions.cs]: Fixed in 9ccbc80. `TakeRandom` now returns the shuffled private copy directly when `count == list.Count` (including the clamped case): `return count == list.Count ? list : list.GetRange(0, count);`. The list is created by `source.ToList()` inside the method, so it is never aliased to caller state and returning it introduces no mutation risk. Verified by the full `EnumerableExtensionsTests` suite (45 tests) on net8.0 and net10.0, and Codex-reviewed (APPROVED).

File: change-log/2026-07-25-252-takerandom-fisher-yates.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# TakeRandom optimised with a partial Fisher-Yates shuffle
+
+**Issue:** [#252](https://github.com/mrploch/ploch-common/issues/252)
+
+## Changed
+
+- `EnumerableExtensions.TakeRandom<T>` now selects items with a partial Fisher-Yates shuffle
+  over an index array instead of repeated `List<T>.RemoveAt` calls, improving complexity from
+  O(count × n) to O(n) setup + O(count) selection. Uniform selection probability and
+  no-duplicate-picks behaviour are unchanged.
```
[the remark this answers is not in the document]
Comment (kploch) [reply change-log/2026-07-25-252-takerandom-fisher-yates.md]: Fixed in 9ccbc80. The change-log entry now describes the actual implementation — a partial Fisher-Yates shuffle performed directly on a private copy of the source list (the "index array" wording was stale from the first iteration of this branch, c1f43e3, before 6f02d1b switched to shuffling the source copy). The PR description has been updated to match as well.

File: src/Common/Collections/EnumerableExtensions.cs [hunk 5]
```diff
@@ -176,37 +176,43 @@ public static IEnumerable<TValue> Shuffle<TValue>(this IEnumerable<TValue> sourc
...
+    /// </param>
     /// <typeparam name="TValue">The enumerable value type.</typeparam>
     /// <returns>The random items from the source enumerable.</returns>
     [SuppressMessage("Security",
                      "CA5394:Do not use insecure randomness",
                      Justification = "Selecting random items from a collection is not a security-sensitive operation; cryptographic randomness is unnecessary and would degrade performance.")]
     public static IEnumerable<TValue> TakeRandom<TValue>(this IEnumerable<TValue> source, int count)
     {
-        var list = source.ToList();
+        if (count <= 0)
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/Collections/EnumerableExtensions.cs:199]: Fixed in c041421. `TakeRandom` now calls `source.NotNull(nameof(source))` before the non-positive-count short-circuit, so a null source always throws `ArgumentNullException` (param name `source`) — restoring the pre-optimisation behaviour and matching `Enumerable.Take`'s eager validation. The behaviour is pinned by a new three-case theory test (`TakeRandom_should_throw_ArgumentNullException_when_source_is_null` with counts -1/0/5), and the `<exception>` XML doc was added. All 45 EnumerableExtensionsTests pass on net8.0 and net10.0; CI is green on the commit.
